Sources/MapboxCoreNavigation/LegacyRouteController.swift

```diff
@@ -375,11 +357,12 @@ open class LegacyRouteController: NSObject, Router, InternalRouter, CLLocationMa
                      RouteController.NotificationUserInfoKey.routingErrorKey: error,
                  ])
                  return
-            case let .success(response):
-                guard case let .route(options) = response.options, !(response.routes?.isEmpty ?? true) else {
-                    return
-                }
-                strongSelf.updateRoute(with: .init(routeResponse: response, routeIndex: 0), routeOptions: options) // unconditionally getting the first route above
+            case let .success(indexedResponse):
+                let response = indexedResponse.routeResponse
+                guard let route = response.routes?[indexedResponse.routeIndex] else { return }
```

Comment:
Should we check array bounds here and in `RouteController`?

Reply:
If `indexedResponse`’s index is out of bounds with respect to its own response, that would be a bug in whatever initialized the IndexedRouteResponse. In this case, it would be `Router.calculateRoutes(from:along:completion:)`, but even ignoring rerouting, the application could’ve passed in an inconsistent IndexedRouteResponse in the first place. It would be reasonable to add a check, but we should do that earlier and more systematically, probably as a separate PR.

Reply:
I added a couple strategic assertions, not as early as I think we probably should to catch developer error, but at least enough to catch internal inconsistency before it ricochets across the SDK. This forced me to revisit one spot where we had been explicitly creating an out-of-bounds IndexedRouteResponse, in the case where the rerouting request came back with no routes. Now that’s an error; everything that kicks off the reroute will abort the reroute.